feat: Add prompt injection protection mechanism

nextedoff opened the pull request:
## Purpose

This feature is adding a way to implement new protection methods for the query, focusing on prompt injection protection in this PR. For injection protection model, model named `protectai/deberta-v3-base-prompt-injection` was used.
New models and guard mechanisms can be added to `promptprotection.py`.

<img width="600" alt="image" src="https://github.com/user-attachments/assets/8335c0f2-1b15-44e3-babf-5c86a070c329">
<img width="600" alt="image" src="https://github.com/user-attachments/assets/df7a489d-a6d1-4f7f-8a8f-875c92208393">

Currently, the injection protection can either be turned on via API and user interface as seen in the screenshot, or it can be set via environment variables:
`USE_INJECTION_PROTECTION="true"`


## Does this introduce a breaking change?

When developers merge from main and run the server, azd up, or azd deploy, will this produce an error?
If you're not sure, try it out on an old environment.

- [ ] Yes
- [x] No

## Does this require changes to learn.microsoft.com docs?

- [ ] Yes
- [x] No

## Type of change

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Code quality checklist

- [ ] The current tests all pass (`python -m pytest`).
- [ ] I added tests that prove my fix is effective or that my feature works
- [x] I ran `python -m pytest --cov` to verify 100% coverage of added lines
- [x] I ran `python -m mypy` to check for type errors
- [x] I either used the pre-commit hooks or ran `ruff` and `black` manually on my code.

nextedoff committed:
Code cleanup

Signed-off-by: nextedoff <rajmukx22@gmail.com>

nextedoff committed:
Implemented a feature to support prompt protection via LLMs, including injection protection

Signed-off-by: nextedoff <rajmukx22@gmail.com>

nextedoff committed:
Added missing files for prompt protection

Signed-off-by: nextedoff <rajmukx22@gmail.com>

nextedoff committed:
Snapshots updates due to the change of error string

Signed-off-by: nextedoff <rajmukx22@gmail.com>

phoevos commented on infra/main.parameters.json:
```diff
@@ -158,6 +158,9 @@
       "value": "${DEFAULT_MODEL}"
     },
+    "useInjectionProtection": {
+      "value": "${USE_INJECTION_PROTECTION}"
```
Let's also add the default here!

```suggestion
      "value": "${USE_INJECTION_PROTECTION=false}"
```

nextedoff replied:
Done

phoevos commented on app/backend/api_wrappers/openai.py:
```diff
@@ -33,6 +33,9 @@ async def chat_completion(self, *args, **kwargs) -> ChatCompletion:
         return await self.client.embeddings.create(*args, **kwargs)
 
+    async def text_classification(self, *args, **kwargs):
+        pass
```
Why don't we raise an error like we do in the HF wrapper?

```suggestion
        raise NotImplementedError
```

What will happen if `check_for_injection` runs with an OpenAI client?

nextedoff replied:
Well, it was not expected that we would do something similar with the OpenAI client, and it's hardcoded to work with the Hugging Face Hub client.

phoevos replied:
Should this be clearer in the code? The OpenAI one supposedly has built-in protections for stuff like that, right?

nextedoff replied:
> Should this be clearer in the code? The OpenAI one supposedly has built-in protections for stuff like that, right?

Yeah, larger models (including gpt3.5-4) usually have guardrails for that.
I will add the notimplemented exception.

nextedoff replied:
Done

phoevos commented on app/backend/api_wrappers/hugging_face.py:
```diff
@@ -74,6 +75,11 @@ async def chat_completion(
         )
 
+    async def text_classification(
+        self, text: str, model: Optional[str] = None
```
Just noticed that we do the same for the chat completion API, but why is the model optional? What happens if we don't provide it?

nextedoff replied:
That's my mistake, I was also planning to make the PromptProtection instance to store its own hugging face client, which would allow to set the default model to use, then this would be optional.

nextedoff replied:
Fixed

phoevos commented on app/backend/app.py:
```diff
@@ -449,6 +452,8 @@ async def setup_clients():
     HUGGINGFACE_API_KEY = os.getenv("HUGGINGFACE_API_KEY")
     DEFAULT_MODEL = os.getenv("DEFAULT_MODEL")
 
+    USE_INJECTION_PROTECTION = os.getenv("USE_INJECTION_PROTECTION", "false").lower() == "true"
```
Isn't it this way?

```suggestion
    USE_INJECTION_PROTECTION = os.getenv("USE_INJECTION_PROTECTION", "").lower() == "true"
```

Not that it won't work, but everywhere else we do it the other way.

nextedoff replied:
fixed

phoevos commented on app/backend/app.py:
```diff
@@ -604,10 +609,15 @@ async def setup_clients():
         hf_client = HuggingFaceClient(token=HUGGINGFACE_API_KEY)
 
     llm_clients: dict[str, LLMClient] = {"openai": openai_client, "hf": hf_client}
+    protection_config = PromptProtectionConfig()
+
+    if USE_INJECTION_PROTECTION:
+        protection_config.injection_protection.use = True
```
This feels a bit wordy. Can't we do something like this?

```suggestion
    protection_config = PromptProtectionConfig(use=USE_INJECTION_PROTECTION)
```

phoevos replied:
Oh no, we can't due to the way these dataclasses are modelled. I'll have to think a bit about this but it does look weird exposing an API like this.

phoevos commented on app/backend/core/promptprotection.py:
```diff
@@ -0,0 +1,58 @@
+
+class PromptProtection:
+    def __init__(self, config: Optional[PromptProtectionConfig]):
+        if config:
+            self.config = config
+        else:
+            self.config = PromptProtectionConfig()
```
```suggestion
        self.config = config if config else PromptProtectionConfig()
```

phoevos commented on app/backend/error.py:
```diff
@@ -7,13 +8,22 @@
 
 def error_dict(error: Exception) -> dict:
-    if isinstance(error, APIError) and error.code == "content_filter":
+    if (isinstance(error, APIError) or isinstance(error, PromptProtectionError)) and error.code == "content_filter":
```
Check out this cool syntactic sugar for what you're doing there (passing a tuple of types to check against)!
```suggestion
    if isinstance(error, (APIError, PromptProtectionError)) and error.code == "content_filter":
```

nextedoff replied:
Done

phoevos commented on app/backend/error.py:
```diff
@@ -22,6 +32,6 @@ def error_dict(error: Exception) -> dict:
 def error_response(error: Exception, route: str, status_code: int = 500):
     logging.exception("Exception in %s: %s", route, error)
-    if isinstance(error, APIError) and error.code == "content_filter":
+    if (isinstance(error, APIError) or isinstance(error, PromptProtectionError)) and error.code == "content_filter":
```
Check out this cool syntactic sugar for what you're doing there (passing a tuple of types to check against)!
```suggestion
    if isinstance(error, (APIError, PromptProtectionError)) and error.code == "content_filter":
```

nextedoff replied:
Done

phoevos reviewed (changes requested):
Thanks for opening this!

Left some comments with suggestions, but I'm not finished yet. I'll continue tomorrow to check the core of the implementation and the frontend changes (which I'm duly excited about)!

nextedoff committed:
Code refactorisation for better handling of multiple protection mechanisms

Signed-off-by: nextedoff <rajmukx22@gmail.com>

nextedoff committed:
Various fixes and improvements

Signed-off-by: nextedoff <rajmukx22@gmail.com>

phoevos commented on app/backend/app.py:
```diff
@@ -605,9 +610,13 @@ async def setup_clients():
     llm_clients: dict[str, LLMClient] = {"openai": openai_client, "hf": hf_client}
 
+    prompt_protection = PromptProtection(prompt_protection_enabled=USE_INJECTION_PROTECTION)
+    print(prompt_protection)
```
Likely a debugging statement, let's drop it!
```suggestion
```

phoevos committed:
fix: Minor docstring changes

Fixed docstrings in promptprotection.py, slightly rephrasing some for
clarity and making sure that they're in proper markdown so that they
are rendered correctly in the documentation.
Renamed the 'config' dictionary of the 'PromptProtection' class to
'protections'.

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Sorry I'm a bloody idiot

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos reviewed (approved):
I'll do some quick manual testing, but other than that LGTM!

phoevos committed:
fix: Use intended settings styling for Ask tab

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Pass checkbox style to ProtectionOptions

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
Merge branch 'main' into feature-injection-model
